### PR TITLE
feat: add governance funding proposal script

### PR DESCRIPTION
**Pull request by Reinis-FRP**
**Motivation**

Automate funding proposals in DVM2

**Summary**

Adds admin proposal script for single token funding from UMA governance.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [x]  Untested

**Issue(s)**

Fixes https://linear.app/uma/issue/UMA-1761/create-and-verify-proposal-for-ostake-uma-transfer

### CODE REVIEW

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    proposerSigner = (await getGckmsSigner()).connect(ethers.provider);
+    if (proposerAddress.toLowerCase() != (await proposerSigner.getAddress()).toLowerCase())
+      throw new Error("GCKMS wallet does not match proposer wallet");
+  } else {
+    if (hre.network.name !== "localhost") throw new Error("Cannot impersonate on mainnet");
+    proposerSigner = await ethers.getImpersonatedSigner(proposerAddress);
+  }
+
+  return proposerSigner;
+}
```
Comment (mrice32) [inline packages/scripts/src/admin-proposals/funding/0_Propose.ts:47]: Do we have a shared common file for these scripts where this function could be implemented and shared?

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  // Verify that GovernorV2 has sufficient balance.
+  const governorV2Address = await getAddress("GovernorV2", 1);
+  const governorBalance = await token.balanceOf(governorV2Address);
+  if (governorBalance.lt(fundingAmount)) throw new Error("Insufficient balance to fund proposal");
+
+  return { token, fundingAmount };
+}
+
+async function verifyProposerBond(params: ProposalParams, proposerV2: ProposerV2Ethers): Promise<void> {
```
Comment (mrice32) [inline packages/scripts/src/admin-proposals/funding/0_Propose.ts:90]: Also seems like a good common utility, no?

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      data: token.interface.encodeFunctionData("transfer", [params.recipient, fundingAmount]),
+    },
+  ];
+  console.log("Sending proposal transactions to the proposer");
+  const ancillaryData = `title: Admin funding proposal, proposalUrl: "${params.proposalUrl}"`;
+  const txn = await proposerV2
+    .connect(params.proposerSigner)
+    .propose(adminProposalTransactions, ethers.utils.toUtf8Bytes(ancillaryData));
+  await txn.wait();
+  if (params.trace) await hre.run("trace", { hash: txn.hash });
```
Comment (mrice32) [inline packages/scripts/src/admin-proposals/funding/0_Propose.ts:130]: What is this? I've never heard of HRE trace? Does this just print out a trace or something?

**Review by mrice32 — APPROVED** [review]
LGTM! Only minor comments

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts (hunk 1 above)
Comment (Reinis-FRP) [reply packages/scripts/src/admin-proposals/funding/0_Propose.ts:47]: Not yet, but agree that we should do this for all admin proposals.
I'll create a separate issue for this

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts (hunk 3 above)
Comment (Reinis-FRP) [reply packages/scripts/src/admin-proposals/funding/0_Propose.ts:130]: yes, it does the same as running `hardhat trace` on command line. I find this helpful to check internal calls and emitted logs

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import { getGckmsSigner } from "@uma/common";
+import { getAddress, ERC20Ethers, ProposerV2Ethers, VotingTokenEthers } from "@uma/contracts-node";
+import { getContractInstance } from "../../utils/contracts";
+import { AdminProposalTransaction } from "../../upgrade-tests/voting2/migrationUtils";
+
+const { ethers } = hre;
+
+require("dotenv").config();
+
+const proposerAddress = "0x2bAaA41d155ad8a4126184950B31F50A1513cE25";
```
Comment (daywiss) [inline packages/scripts/src/admin-proposals/funding/0_Propose.ts:23]: interesting this is not an env

**Review by daywiss — APPROVED** [review]
nice

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts (hunk 1 above)
Comment (mrice32) [reply packages/scripts/src/admin-proposals/funding/0_Propose.ts:47]: Perfect!

**Review by mrice32 — APPROVED** [review]

File: packages/scripts/src/admin-proposals/funding/0_Propose.ts (hunk 4 above)
Comment (Reinis-FRP) [reply packages/scripts/src/admin-proposals/funding/0_Propose.ts:23]: agree, this is not optimal. better to use env when simulating and derive from keys when doing real proposal. Could improve this when refactoring getting proposer in a common lib for all admin proposals as per https://linear.app/uma/issue/UMA-1773/refactor-admin-proposal-scripts-to-share-common-functions

**Pull request merged**